kore/src/SMT.hs

```diff
@@ -517,11 +529,18 @@ stopSolver mvar = do
 
 -- | Run an external SMT solver.
 runSMT :: Config -> SMT () -> SMT a -> LoggerT IO a
-runSMT config userInit smt =
-    Exception.bracket
-        (newSolver config)
-        stopSolver
-        (\mvar -> runSMT' config userInit mvar smt)
+runSMT config userInit smt = case autoRestart config of
+    Never -> smtBracket
+    Once ->
+        smtBracket `catch` \(e :: SimpleSMT.SolverException) -> do
+            warnZ3Crash e
+            smtBracket
```

Comment:
I'm confused, won't this basically re-run the whole prover?

I mean, what we want is if during some query we find that Z3 has crashed, we restart _just_ Z3 and retry that specific query, resuming execution from that point.